assembly/BytesReader.ts

```diff
@@ -0,0 +1,169 @@
+        this.index += size;
+        return bytes;
+    }
+
+    /**
+     * Read bytes into Int8
+     */
+    readInt8(): Int8 {
+        const value = Int8.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into Int16
+     */
+    readInt16(): Int16 {
+        const value = Int16.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into Int32
+     */
+    readInt32(): Int32 {
+        const value = Int32.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into Int64
+     */
+    readInt64(): Int64 {
+        const value = Int64.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into UInt8
+     */
+    readUInt8(): UInt8 {
+        const value = UInt8.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();        
+        return value;
+    }
+    /**
+     * Read bytes into UInt16
+     */
+    readUInt16(): UInt16 {
+        const value = UInt16.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into UInt32
+     */
+    readUInt32(): UInt32 {
+        const value = UInt32.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into UInt64
+     */
+    readUInt64(): UInt64 {
+        const value = UInt64.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into CompactInt
+     */
+    readCompactInt(): CompactInt{
+        const value = CompactInt.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+
+    /**
+     * Read byte into Bool
+     */
+    readBool(): Bool {
+        const value = Bool.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into Hash
+     */
+    readHash(): Hash {
+        const value = Hash.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into ByteArray
+     */
+    readByteArray(): ByteArray {
+        const value = ByteArray.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Read bytes into ScaleString
+     */
+    readScaleString(): ScaleString {
+        const value = ScaleString.fromU8a(this.bytes, this.index);
+        this.index += value.encodedLength();
+        return value;
+    }
+    /**
+     * Returns leftover bytes after reading bytes array
```

Comment:
`Returns leftover bytes after reading bytes array` - it is not necessary that this will be called "after reading bytes array"..
IMO a better description would be: `Returns the unread bytes from the reader`